File: 14/solution.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"log"
	"os"
	"regexp"
	"strconv"
)

func setBit(value uint64, pos uint) uint64 {
	value |= (1 << pos)
	return value
}

func clearBit(value uint64, pos uint) uint64 {
	value &= ^(1 << pos)
	return value
}

func applyMaskToVal(inputInt uint64, mask string) uint64 {
	for i := len(mask) - 1; i >= 0; i-- {
		pos := uint(len(mask) - i - 1)
		if string(mask[i]) == "1" {
			inputInt = setBit(inputInt, pos)
		} else if string(mask[i]) == "0" {
			inputInt = clearBit(inputInt, pos)
		}
	}
	return inputInt
}

func part1(lines []string) uint64 {
	var mask string
	memory := make(map[uint64]uint64)
	expression := regexp.MustCompile(`\d+`)
	for _, line := range lines {
		if line[:4] == "mask" {
			mask = line[7:]
		} else {
			match := expression.FindAllString(line, -1)
			addr, err := strconv.ParseUint(match[0], 10, 64)
			if err != nil {
				log.Fatal(err)
			}
			value, err := strconv.ParseUint(match[1], 10, 64)
			if err != nil {
				log.Fatal(err)
			}

			memory[addr] = applyMaskToVal(value, mask)
		}
	}
	sum := uint64(0)
	for _, v := range memory {
		sum += v
	}
	return sum
}

func applyMaskToAddr(inputInt uint64, mask string) []uint64 {
	result := make(map[uint64]struct{})
	for i := len(mask) - 1; i >= 0; i-- {
		pos := uint(len(mask) - i - 1)
		if string(mask[i]) == "1" {
			inputInt = setBit(inputInt, pos)
		}
	}
	result[inputInt] = struct{}{}
	for i := len(mask) - 1; i >= 0; i-- {
		pos := uint(len(mask) - i - 1)
		if string(mask[i]) == "X" {
			var temp []uint64
			for addr := range result {
				temp = append(temp, addr)
			}
			for _, addr := range temp {
				result[setBit(addr, pos)] = struct{}{}
				result[clearBit(addr, pos)] = struct{}{}
			}
		}
	}
	var resultSlice []uint64
	for k := range result {
		resultSlice = append(resultSlice, k)
	}
	return resultSlice
}

func part2(lines []string) uint64 {
	var mask string
	memory := make(map[uint64]uint64)
	expression := regexp.MustCompile(`\d+`)
	for _, line := range lines {
		if line[:4] == "mask" {
			mask = line[7:]
		} else {
			match := expression.FindAllString(line, -1)
			addr, err := strconv.ParseUint(match[0], 10, 64)
			if err != nil {
				log.Fatal(err)
			}
			value, err := strconv.ParseUint(match[1], 10, 64)
			if err != nil {
				log.Fatal(err)
			}
			addresses := applyMaskToAddr(addr, mask)
			for _, maskedAddr := range addresses {
				memory[maskedAddr] = value
			}

		}
	}
	sum := uint64(0)
	for _, v := range memory {
		sum += v
	}
	return sum
}

func main() {
	flag.Parse()
	if len(flag.Args()) == 0 {
		log.Fatal("Provide the path to input")
	}
	file, err := os.Open(flag.Args()[0])
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	var lines []string
	for scanner.Scan() {
		line := scanner.Text()
		lines = append(lines, line)
	}

	fmt.Printf("Part 1 answer:%v\nPart 2 answer:%v\n", part1(lines), part2(lines))
}

File: 14/solution.py
import sys
import re
import math


def set_bit(value, pos):
    return value | (1 << pos)


def clear_bit(value, pos):
    return value & ~(1 << pos)


def apply_mask_to_val(input_int, mask):
    for i in range(len(mask)-1, -1, -1):
        pos = len(mask) - i - 1
        if mask[i] == '1':
            input_int = set_bit(input_int, pos)
        elif mask[i] == '0':
            input_int = clear_bit(input_int, pos)
    return input_int


def part1(lines):
    mask = ''
    memory = {}
    expression = re.compile(r"\d+")
    for line in lines:
        if line[:4] == "mask":
            mask = line[7:-1]
        else:
            match = expression.findall(line)
            memory[int(match[0])] = apply_mask_to_val(int(match[1]), mask)
    return sum(memory.values())


def apply_mask_to_addr(input_int, mask):
    result = set()
    for i in range(len(mask)-1, -1, -1):
        pos = len(mask) - i - 1
        if mask[i] == '1':
            input_int = set_bit(input_int, pos)
    result.add(input_int)
    for i in range(len(mask)-1, -1, -1):
        pos = len(mask) - i - 1
        if mask[i] == 'X':
            temp = list(result)
            for addr in temp:
                result.add(set_bit(addr, pos))
                result.add(clear_bit(addr, pos))
    return result


def part2(lines):
    mask = ''
    memory = {}
    expression = re.compile(r"\d+")
    for line in lines:
        if line[:4] == "mask":
            mask = line[7:-1]
        else:
            match = expression.findall(line)
            addresses = apply_mask_to_addr(int(match[0]), mask)
            for masked_address in addresses:
                memory[masked_address] = int(match[1])

    return sum(memory.values())


try:
    f = open(sys.argv[1], "r")
except IOError:
    print("Error opening the file, try again")
    sys.exit(1)
with f:
    rawLines = f.readlines()
    f.close()
    # digitArray = [int(s) for s in lines]
    print(
        f"Part 1 answer:{part1(rawLines)} Part 2 answer: {part2(rawLines)}")
